text/2540-formal-function-parameter-attributes.md

```diff
@@ -0,0 +1,242 @@
+Please note that I am *not* proposing to actually add this to the language!
+
+## Example: Procedural Macros
+
+Also procedural macros could greatly benefit from having their own defined custom attributes on formal parameters.
+
+# Guide-level explanation
+[guide-level-explanation]: #guide-level-explanation
+
+Formal parameters of functions, methods, closures and functions in trait definitions may have attributes attached to them.
+This allows to provide additional information to a given formal parameter.
+
+To reify this we introduce the `#[unused]` attribute that states that the attributed parameter is unused in the associated implementation.
+
+## Examples
+
+The syntax for this is demonstrated by the code below:
+
+```rust
+// Function
+fn foo(#[unused] bar: u32) { .. }
+
+// Methods & trait & definitions:
+// - `self` can also be attributed
+fn foo(#[unused] self, ..) { .. }
+fn foo(#[unused] &self, ..) { .. }
+fn foo(#[unused] &mut self, ..) { .. }
+
+// Closures & Lamdas
+|#[unused] x| { .. }
+```
+
+### Trait declarations
+
+```rust
+fn foo(#[unused] self);
+```
+
+Note that while the `#[unused]` attribute is syntactically
+possible to put here it doesn't actually make sense semantically
+since method declarations have no implementation.
+Other attributes might be very useful as for formal parameters in a method declaration.
+
+## Errors & Warnings
+
+There are two errornous situations when it comes to 
+
+### Warning: Unused attribute
+
+When using an non-defined attribute that is not used by either the language or a custom defined procedural macro.
+
+```
+warning: unused attribute
+ --> src/main.rs:2
+  |
+2 | #[foo] bar: u32
+  | ^^^^^^^^^
+  |
+  = note: #[warn(unused_attributes)] on by default
+```
+
+### Error: Malformed attribute
+
+When using a known attribute that is not defined for formal parameters such as when attributing `fn main` with `#[allow]`.
+
+Example shows the usage of the known attribute `#[inline]` used on a formal parameter without being defined for it.
+
+```
+error[E0452]: malformed lint attribute
+ --> src/main.rs:2
+  |
+2 | #[inline] bar: u32
+  | ^^^^^^^^
+```
+
+The same applies for attributes with an incorrect format such as `#[inline(key = value)]` that is handled as its done in other contexts.
+
+
+# Reference-level explanation
+[reference-level-explanation]: #reference-level-explanation
+
+## Description
+
+In accordance to the RFC for [attributes for generic params](https://github.com/frol/rust-rfcs/blob/master/text/1327-dropck-param-eyepatch.md)
+this feature is guarded by the `formal_function_param_attrs` feature guard.
+
+The grammar of the following language items has to be adjusted to allow that
+constructions like the following will become legal.
+
+- Function definitions
+- Method definitions
+- Trait function declarations and defnitions (with or without default impl)
+- Lambda & closure definitions
```

Comment:
Lambdas and closures are the same thing.

Reply:
I always thought that closures are stateful while lambdas aren't.

Reply:
"Closure" is more a word used to describe an implementation method or a word that is used more in non-functional programming languages. But they are the same word.

For example, we may define the following lambda (untyped lambda calculus):
```
λ x. λ y. x
```

or in Rust:

```rust
let const = |x: u8| move |y: u8| x;
```

The inner lambda captures ("stateful") the outer binding `x`.

Reply:
Okay then I will correct all usages in the RFC.